File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/data/model/InternalUser.java
package com.github.arhor.simple.expense.tracker.data.model;

import lombok.Builder;

import java.time.LocalDateTime;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Immutable;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.relational.core.mapping.Table;

@Table("users")
@Builder(toBuilder = true)
@Immutable
public record InternalUser(
    @Id
    Long id,
    String username,
    String password,
    String currency,
    String externalId,
    String externalProvider,
    @CreatedDate
    LocalDateTime created,
    @LastModifiedDate
    LocalDateTime updated
) {

    /**
     * Compact projection of the {@link InternalUser} entity. Main purpose is to load only necessary fields from DB.
     */
    public record CompactProjection(Long id, String username, String currency) {
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/data/model/InternalUser.kt
package com.github.arhor.simple.expense.tracker.data.model;

import lombok.Builder
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Immutable
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.relational.core.mapping.Table
import java.time.LocalDateTime

@Table("users")
@Builder(toBuilder = true)
@Immutable
data class InternalUser(
    @Id
    val id: Long? = null,
    val username: String? = null,
    val password: String? = null,
    val currency: String,
    val externalId: String? = null,
    val externalProvider: String? = null,
    @CreatedDate
    val created: LocalDateTime? = null,
    @LastModifiedDate
    val updated: LocalDateTime? = null,
)
